File: 알고리즘 스터디/개인공부/Greedy/LinkedIsland.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42861

import sys
import heapq


def get_parent(node, parent) -> int:
    if parent[node] == node:
        return parent[node]

    return get_parent(parent[node], parent)


def solution(n, costs):
    answer = 0
    parent = [i for i in range(n)]
    path = [dict() for _ in range(n)]
    queue = []

    for _src, _dst, _cost in costs:
        path[_src][_dst] = min(path[_src].get(_dst, sys.maxsize), _cost)
        path[_dst][_src] = min(path[_dst].get(_src, sys.maxsize), _cost)

        heapq.heappush(queue, (_cost, _src, _dst))

    while queue:
        _cost, _src, _dst = heapq.heappop(queue)

        p1, p2 = get_parent(_src, parent), get_parent(_dst, parent)

        # 부모노드 체크
        if p1 == p2:
            continue

        parent[p1] = parent[p2]

        answer += _cost

    return answer


if __name__ == '__main__':
    solution(4, [[0, 1, 1], [0, 2, 2], [1, 2, 5], [1, 3, 1], [2, 3, 8]])  # 4

File: 알고리즘 스터디/개인공부/Greedy/LinkedIsland.java
/* https://school.programmers.co.kr/learn/courses/30/lessons/42861 */

import java.util.Arrays;
import java.util.Comparator;

public class LinkedIsland {
    class Island {
        int cost, island1, island2;

        public Island(int island1, int island2, int cost) {
            this.cost = cost;
            this.island1 = island1;
            this.island2 = island2;
        }
    }

    public int solution(final int n, final int[][] costs) {
        int answer = 0;
        PriorityQueue<Island> pq = new PriorityQueue<>((o1, o2) -> o1.cost - o2.cost);
        Map<Integer, LinkedList<Island>> path = new HashMap<>(n);
        boolean[] visited = new boolean[n];

        for (int[] cost : costs) {
            int[] _cost = cost;

            if (!path.containsKey(_cost[0]))
                path.put(new Integer(_cost[0]), new LinkedList<>());

            if (!path.containsKey(_cost[1]))
                path.put(new Integer(_cost[1]), new LinkedList<>());

            path.get(new Integer(_cost[0])).add(new Island(_cost[0], _cost[1], _cost[2]));
            path.get(new Integer(_cost[1])).add(new Island(_cost[1], _cost[0], _cost[2]));
        }

        pq.offer(new Island(0, 0, 0));

        Arrays.fill(visited, false);

        while (!pq.isEmpty()) {
            Island cur_i = pq.poll();

            if (visited[cur_i.island2])
                continue;

            visited[cur_i.island2] = true;
            answer += cur_i.cost;

            for (Island i : path.get(cur_i.island2)) {
                if (visited[i.island2])
                    continue;

                pq.offer(i);
            }
        }

        return answer;
    }

}
